From: Jeff Moyer <jmoyer@redhat.com>
Date: Tue, 8 Dec 2009 18:58:02 -0500
Subject: [block] blkio: Set must_dispatch only if we decided to not dispatch the request
Message-id: <1260298712-12756-30-git-send-email-jmoyer@redhat.com>
Patchwork-id: 21775
O-Subject: [RHEL6 PATCH 29/59] blkio: Set must_dispatch only if we decided to
	not dispatch the request
Bugzilla: 425895
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

fixes bug 425895

commit bf7919371025412978268efca4b09dd847acb395
Author: Vivek Goyal <vgoyal@redhat.com>
Date:   Thu Dec 3 12:59:37 2009 -0500

    blkio: Set must_dispatch only if we decided to not dispatch the request

    o must_dispatch flag should be set only if we decided not to run the queue
      and dispatch the request.

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 8743057..c5fa26f 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -2489,9 +2489,9 @@ cfq_rq_enqueued(struct cfq_data *cfqd, struct cfq_queue *cfqq,
 			if (blk_rq_bytes(rq) > PAGE_CACHE_SIZE ||
 			    cfqd->busy_queues > 1) {
 				del_timer(&cfqd->idle_slice_timer);
-			__blk_run_queue(cfqd->queue);
-			}
-			cfq_mark_cfqq_must_dispatch(cfqq);
+				__blk_run_queue(cfqd->queue);
+			} else
+				cfq_mark_cfqq_must_dispatch(cfqq);
 		}
 	} else if (cfq_should_preempt(cfqd, cfqq, rq)) {
 		/*
